pkg/provider/aws/resources/eks.go

```diff
@@ -288,35 +289,75 @@ func NodeGroupName(networkPlacement string, instanceType string) string {
 	return nodeGroupSanitizer.Apply(fmt.Sprintf("%s_%s", networkPlacement, instanceType))
 }
 
-func createAddOns(cluster *EksCluster, provenance []core.AnnotationKey) []*kubernetes.HelmChart {
-	return []*kubernetes.HelmChart{
-		{
-			Name:          cluster.Name + `-metrics-server`,
-			Chart:         "metrics-server",
-			ConstructRefs: provenance,
-			ClustersProvider: core.IaCValue{
-				Resource: cluster.Kubeconfig,
-				Property: CLUSTER_PROVIDER_IAC_VALUE,
-			},
-			Repo: `https://kubernetes-sigs.github.io/metrics-server/`,
+func createAddOns(cluster *EksCluster, vpc *Vpc, provenance []core.AnnotationKey) []core.Resource {
+	clustersProvider := core.IaCValue{
+		Resource: cluster,
+		Property: CLUSTER_PROVIDER_IAC_VALUE,
+	}
+	return []core.Resource{
+		&kubernetes.HelmChart{
+			Name:             cluster.Name + `-metrics-server`,
+			Chart:            "metrics-server",
+			ConstructRefs:    provenance,
+			ClustersProvider: clustersProvider,
+			Repo:             `https://kubernetes-sigs.github.io/metrics-server/`,
 		},
-		{
-			Name:          cluster.Name + `-cert-manager`,
-			Chart:         `cert-manager`,
-			ConstructRefs: provenance,
-			ClustersProvider: core.IaCValue{
-				Resource: cluster.Kubeconfig,
-				Property: CLUSTER_PROVIDER_IAC_VALUE,
-			},
-			Repo:    `https://charts.jetstack.io`,
-			Version: `v1.10.0`,
+		&kubernetes.HelmChart{
+			Name:             cluster.Name + `-cert-manager`,
+			Chart:            `cert-manager`,
+			ConstructRefs:    provenance,
+			ClustersProvider: clustersProvider,
+			Repo:             `https://charts.jetstack.io`,
+			Version:          `v1.10.0`,
 			Values: map[string]any{
 				`installCRDs`: true,
 				`webhook`: map[string]any{
 					`timeoutSeconds`: 30,
 				},
 			},
 		},
+		&HelmChartAlbController{
+			HelmChart: kubernetes.HelmChart{
+
+				Name:             cluster.Name + `-alb-c`,
+				Chart:            `aws-load-balancer-controller`,
+				ConstructRefs:    provenance,
+				Repo:             `https://aws.github.io/eks-charts`,
+				Version:          `1.4.7`,
+				ClustersProvider: clustersProvider,
+				Namespace:        "default", // kube-system if on fargate, default otherwise
+			},
+			ClusterName: cluster.Name,
+			Region:      NewRegion(),
+			Vpc:         vpc,
+			Role:        NewIamRole(cluster.Name, `alb-controller-role`, provenance, ServiceRolePolicyDocument(cluster, `alb-controller`)),
```

Comment:
this role needs a specific policy for it to be able to work. we have all the policy defined in deploylib but its a custom one we have to write

Reply:
Is that policy not what this sets up, via `ServiceRolePolicyDocument`?

The deploylib (`eks.ts`, more precisely) code is:

```typescript
const assumeRolePolicyDoc = pulumi
    .all([this.oidcProvider.arn, this.oidcProvider.url])
    .apply(([oidc_arn, oidc_url]) => {
        return {
            Version: '2012-10-17',
            Statement: [
                {
                    Effect: 'Allow',
                    Principal: {
                        Federated: oidc_arn,
                    },
                    Action: 'sts:AssumeRoleWithWebIdentity',
                    Condition: {
                        StringEquals: {
                            [`${oidc_url}:sub`]: `system:serviceaccount:${namespace}:${name}`,
                        },
                    },
                },
            ],
        }
    })
```

the new `index.ts` we generate, after prettying, is:

```typescript
 pulumi
  .all([
    awsEksClusterTsEksHelmEksCluster.openIdConnectIssuerUrl,
    awsEksClusterTsEksHelmEksCluster.arn,
  ])
  .apply(([cluster_oidc_url, cluster_arn]) => {
    return {
      Version: `2013-10-17`,
      Statement: [
        {
          Effect: `Allow`,
          Action: [`sts:AssumeRoleWithWebIdentity`],
          Principal: {
            Federated: `arn:aws:iam::${
              cluster_arn.split(":")[5]
            }:oidc-provider/${cluster_oidc_url}`,
          },
          Condition: {
            StringEquals: {
              [`${cluster_oidc_url}:sub`]: `system:serviceaccount:default:alb-controller`,
            },
          },
        },
      ],
    };
  })
```

They seem pretty much the same, except for the `oidc_arn`, which is `this.oidcProvider.arn` in the deployib version and `'arn:aws:iam::${cluster_arn.split(":")[5]}:oidc-provider/${cluster_oidc_url}'` in the new. Would that evaluate to something different?

Reply:
sorry, referencing https://github.com/klothoplatform/klotho/blob/main/pkg/infra/pulumi_aws/iac/k8s/add_ons/alb_controller/index.ts#L38 as in the permissions it needs to call the ec2 api, etc, to spin up the resources for load balancers and other aws resources.

This policy should be good for assuming the role though